File: app/src/main/java/vn/loitp/app/activity/database/realm/RealmActivity.java
package vn.loitp.app.activity.database.realm;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;

import java.util.List;

import io.realm.Realm;
import io.realm.RealmResults;
import loitp.basemaster.R;
import vn.loitp.app.activity.demo.ebookwithrealm.EbookWithRealmActivity;
import vn.loitp.core.base.BaseActivity;
import vn.loitp.core.utilities.LActivityUtil;
import vn.loitp.core.utilities.LLog;

public class RealmActivity extends BaseActivity implements View.OnClickListener {
    private Realm mRealm;
    private LinearLayout ll;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mRealm = RealmController.with(getApplication()).getRealm();
        ll = (LinearLayout) findViewById(R.id.ll);
        findViewById(R.id.bt_realm).setOnClickListener(this);
        findViewById(R.id.bt_add).setOnClickListener(this);
        findViewById(R.id.bt_clear_all).setOnClickListener(this);

        // refresh the realm instance
        RealmController.with(getApplication()).refresh();

        getAll();
    }

    @Override
    protected boolean setFullScreen() {
        return false;
    }

    @Override
    protected String setTag() {
        return getClass().getSimpleName();
    }

    @Override
    protected int setLayoutResourceId() {
        return R.layout.activity_realm;
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.bt_realm:
                Intent intent = new Intent(activity, EbookWithRealmActivity.class);
                startActivity(intent);
                LActivityUtil.tranIn(activity);
                break;
            case R.id.bt_add:
                add();
                break;
            case R.id.bt_clear_all:
                clearUI();
                RealmController.getInstance().clearAllMyBook();
                getAll();
                break;
        }
    }

    private void add() {
        LLog.d(TAG, "add " + RealmController.getInstance().getBooks().size());
        mRealm.beginTransaction();

        MyBook myBook = new MyBook();
        myBook.setId(RealmController.getInstance().getBooks().size() + System.currentTimeMillis());
        myBook.setTitle(RealmController.getInstance().getBooks().size() + " title");

        //logMyBook(myBook);

        mRealm.copyToRealm(myBook);
        mRealm.commitTransaction();

        addButton(myBook);
    }

    private void logMyBook(MyBook myBook) {
        LLog.d(TAG, "_______________________________________");
        LLog.d(TAG, ">>> " + myBook.getId() + " - " + myBook.getTitle());
    }

    private void logMyBook(List<MyBook> myBookList) {
        LLog.d(TAG, "_______________________________________");
        for (MyBook mb : myBookList) {
            LLog.d(TAG, ">>> " + mb.getId() + " - " + mb.getTitle());
        }
    }

    private void addButton(MyBook myBook) {
        Button button = new Button(activity);
        button.setText(myBook.getTitle() + " - " + myBook.getId());
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                clickMyBook(myBook, button);
            }
        });
        button.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                longClickMyBook(myBook, button);
                return true;
            }
        });
        ll.addView(button);
    }

    private void clearUI() {
        List<View> viewList = ll.getTouchables();
        for (View view : viewList) {
            if (view instanceof Button) {
                ll.removeView(view);
            }
        }
    }

    private void printUI(List<MyBook> myBookList) {
        for (MyBook mb : myBookList) {
            addButton(mb);
        }
    }

    private void getAll() {
        List<MyBook> myBookList = RealmController.getInstance().getMyBookListSortByID();
        LLog.d(TAG, "getAll: " + myBookList.size());
        //logMyBook(myBookList);
        printUI(myBookList);
    }

    private void clickMyBook(MyBook myBook, Button button) {
        LLog.d(TAG, "clickMyBook");
        //logMyBook(myBook);
        remove(myBook, button);
    }

    private void remove(MyBook myBook, Button button) {
        mRealm.beginTransaction();
        RealmResults<MyBook> myBookRealmResults = RealmController.getInstance().getBooks(myBook);
        if (!myBookRealmResults.isEmpty()) {
            for (int i = myBookRealmResults.size() - 1; i >= 0; i--) {
                myBookRealmResults.get(i).removeFromRealm();
            }
        }
        mRealm.commitTransaction();
        ll.removeView(button);
    }

    private void longClickMyBook(MyBook myBook, Button button) {
        LLog.d(TAG, "longClickMyBook " + button.getText().toString());

        mRealm.beginTransaction();

        MyBook mb = RealmController.getInstance().getMyBook(myBook.getId());
        mb.setTitle("UPDATED " + mb.getTitle());

        mRealm.commitTransaction();

        button.setText(myBook.getTitle() + " - " + myBook.getId());
        LLog.d(TAG, "longClickMyBook done");
    }
}

File: app/src/main/java/vn/loitp/app/activity/database/realm/RealmActivity.kt
package vn.loitp.app.activity.database.realm

import android.annotation.SuppressLint
import android.app.AlertDialog
import android.content.Intent
import android.os.Bundle
import android.widget.Button
import android.widget.EditText
import com.loitpcore.annotation.IsFullScreen
import com.loitpcore.annotation.LogTag
import com.loitpcore.core.base.BaseFontActivity
import com.loitpcore.core.utilities.LActivityUtil
import com.loitpcore.views.setSafeOnClickListener
import io.realm.Realm
import kotlinx.android.synthetic.main.activity_realm.*
import vn.loitp.app.R
import vn.loitp.app.activity.demo.ebookwithrealm.EbookWithRealmActivity

@LogTag("RealmActivity")
@IsFullScreen(false)
class RealmActivity : BaseFontActivity() {
    private lateinit var mRealm: Realm

    override fun setLayoutResourceId(): Int {
        return R.layout.activity_realm
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        mRealm = RealmController.with(application).realm

        // refresh the realm instance
        RealmController.with(application).refresh()

        getAllMyBook()
        setupViews()
    }

    private fun setupViews() {
        btRealm.setSafeOnClickListener {
            val intent = Intent(this, EbookWithRealmActivity::class.java)
            startActivity(intent)
            LActivityUtil.tranIn(this)
        }
        btAdd.setSafeOnClickListener {
            addMyBook()
        }
        btClearAll.setSafeOnClickListener {
            clearUI()
            RealmController.getInstance(this).clearAllMyBook()
            getAllMyBook()
        }
        btSearchById.setSafeOnClickListener {
            searchById()
        }
    }

    private fun addMyBook() {
        mRealm.beginTransaction()

        val myBook = MyBook()
        myBook.id = System.currentTimeMillis()
        myBook.title = "Title ${System.currentTimeMillis()}"
        myBook.author = "Author ${System.currentTimeMillis()}"

        mRealm.copyToRealm(myBook)
        mRealm.commitTransaction()

        addViewMyBook(myBook)
        printTotalSize()
    }

    @SuppressLint("SetTextI18n")
    private fun addViewMyBook(myBook: MyBook) {
        val button = Button(this)
        button.text = "${myBook.title} - ${myBook.author}"
        button.textSize = 10f
        button.isAllCaps = false
        button.setOnClickListener { updateMyBook(myBook, button) }
        button.setOnLongClickListener {
            removeMyBook(myBook, button)
            true
        }
        ll.addView(button)
    }

    private fun clearUI() {
        val viewList = ll.touchables
        for (view in viewList) {
            if (view is Button) {
                ll.removeView(view)
            }
        }
    }

    private fun printUI(myBookList: List<MyBook>) {
        for (mb in myBookList) {
            addViewMyBook(mb)
        }
    }

    private fun getAllMyBook() {
        val myBookList = RealmController.getInstance(this).myBookListSortByID
        printUI(myBookList)
        printTotalSize()
    }

    private fun removeMyBook(myBook: MyBook, button: Button) {
        mRealm.beginTransaction()
        val mb = RealmController.getInstance(this).getMyBook(myBook)
        if (!mb.isEmpty()) {
            for (i in mb.indices.reversed()) {
                mb[i].removeFromRealm()
            }
        }
        mRealm.commitTransaction()
        ll.removeView(button)
        printTotalSize()
    }

    @SuppressLint("SetTextI18n")
    private fun updateMyBook(myBook: MyBook, button: Button) {
        mRealm.beginTransaction()
        val mb = RealmController.getInstance(this).getMyBook(myBook.id)
        mb.title = "Title ${System.currentTimeMillis()}"
        mb.author = "Author ${System.currentTimeMillis()}"
        mRealm.commitTransaction()
        button.text = "${mb.title} - ${mb.author}"
    }

    private fun printTotalSize() {
        showShortInformation("Total size: " + RealmController.getInstance(this).myBookList.size)
    }

    private fun showInputDialog(ipb: InputCallback) {
        val builder = AlertDialog.Builder(this)
        builder.setTitle("Title")

        val input = EditText(this)
        // input.inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_PASSWORD
        builder.setView(input)

        builder.setPositiveButton("OK") { dialog, _ ->
            val text = input.text.toString()
            ipb.onText(text)
            dialog.cancel()
        }
        val dialog = builder.create()
        dialog.show()
    }

    private fun searchById() {
        showInputDialog(object : InputCallback {
            override fun onText(text: String?) {
                text?.let {
                    logD("searchById $it")
                    try {
                        val id: Long = it.toLong()
                        val mb = RealmController.getInstance(this@RealmActivity).getMyBook(id)
                        showShortInformation("searchById -> ${mb.title}")
                    } catch (e: Exception) {
                        showShortError("searchById $e")
                    }
                }
            }
        })
    }
}
